selftests/functional/test_interrupt.py

```diff
@@ -67,64 +77,87 @@ def test_badly_behaved(self):
-                                          'seconds for test to finish '
-                                          '(ignoring new Ctrl+C until then)')
-        # We have to actually wait 2 seconds until the ignore window is over
+        cmd = ('%s run %s --sysinfo=off --job-results-dir %s ' %
+               (AVOCADO, self.test_module, self.tmpdir))
+        self.proc = subprocess.Popen(cmd.split(),
+                                     stdout=subprocess.PIPE,
+                                     stderr=subprocess.STDOUT)
+
+        if not wait.wait_for(self._has_children, timeout=2):
+            process.kill_process_tree(self.proc.pid)
+            self.fail('Avocado did not start the test process.')
+
+        # This test will ignore SIGINT, so it should terminate
+        # when we send the second SIGINT.
+        os.kill(self.proc.pid, signal.SIGINT)
+        # We have to actually wait 2+ seconds until
+        # the ignore window is over
         time.sleep(2.5)
-        proc.sendline('\x03')
-        proc.read_until_last_line_matches('JOB TIME   : %d s')
-        wait.wait_for(lambda: not proc.is_alive(), timeout=1)
+        os.kill(self.proc.pid, signal.SIGINT)
 
-        # Make sure the bad test will be really gone from the process table
-        def wait_until_no_badtest():
-            bad_test_processes = []
+        if not wait.wait_for(self._is_finished, timeout=2):
+            process.kill_process_tree(self.proc.pid)
+            self.fail('Avocado was still running after receiving SIGINT '
+                      'twice.')
 
-            old_psutil = False
-            try:
-                process_list = psutil.pids()
-            except AttributeError:
-                process_list = psutil.get_pid_list()
-                old_psutil = True
-
-            for p in process_list:
-                try:
-                    p_obj = psutil.Process(p)
-                    if p_obj is not None:
-                        if old_psutil:
-                            cmdline_list = psutil.Process(p).cmdline
-                        else:
-                            try:
-                                cmdline_list = psutil.Process(p).cmdline()
-                            except psutil.AccessDenied:
-                                cmdline_list = []
-                        if bad_test.path in " ".join(cmdline_list):
-                            bad_test_processes.append(p_obj)
-                # psutil.NoSuchProcess happens when the original
-                # process already ended and left the process table
-                except psutil.NoSuchProcess:
-                    pass
-
-            return len(bad_test_processes) == 0
-
-        wait.wait_for(wait_until_no_badtest, timeout=2)
+        if not wait.wait_for(self._no_test_in_process_table, timeout=2):
+            self.fail('Avocado left processes behind.')
+
+        output = self.proc.communicate()[0]
```

Comment:
This works, but I think it passes the wrong message to readers (it happened to me, at least).  Popen.communicate has the side effect of returning stdout (and stderr), but it's intended to send data to the process' stdin.  How about:

```diff
diff --git a/selftests/functional/test_interrupt.py b/selftests/functional/test_interrupt.py
index 4325b1a3..e9d9ce31 100644
--- a/selftests/functional/test_interrupt.py
+++ b/selftests/functional/test_interrupt.py
@@ -106,7 +106,7 @@ class InterruptTest(unittest.TestCase):
         if not wait.wait_for(self._no_test_in_process_table, timeout=2):
             self.fail('Avocado left processes behind.')
 
-        output = self.proc.communicate()[0]
+        output = self.proc.stdout.read()
         # Make sure the Interrupted requested sentence is there
         self.assertIn('Interrupt requested. Waiting 2 seconds for test to '
                       'finish (ignoring new Ctrl+C until then)', output)
@@ -150,9 +150,8 @@ class InterruptTest(unittest.TestCase):
         if not wait.wait_for(self._no_test_in_process_table, timeout=2):
             self.fail('Avocado left processes behind.')
 
-        output = self.proc.communicate()[0]
         # Make sure the Interrupted test sentence is there
-        self.assertIn('Terminated\n', output)
+        self.assertIn('Terminated\n', self.proc.stdout.read())
 
     @unittest.skipIf(int(os.environ.get("AVOCADO_CHECK_LEVEL", 0)) < 1,
                      "Skipping test that take a long time to run, are "
@@ -191,7 +190,7 @@ class InterruptTest(unittest.TestCase):
         if not wait.wait_for(self._no_test_in_process_table, timeout=2):
             self.fail('Avocado left processes behind.')
 
-        output = self.proc.communicate()[0]
+        output = self.proc.stdout.read()
         # Make sure the Interrupted requested sentence is there
         self.assertIn('Interrupt requested. Waiting 2 seconds for test to '
                       'finish (ignoring new Ctrl+C until then)', output)
@@ -235,9 +234,8 @@ class InterruptTest(unittest.TestCase):
         if not wait.wait_for(self._no_test_in_process_table, timeout=2):
             self.fail('Avocado left processes behind.')
 
-        output = self.proc.communicate()[0]
         # Make sure the Interrupted test sentence is there
-        self.assertIn('Terminated\n', output)
+        self.assertIn('Terminated\n', self.proc.stdout.read())
 
     def tearDown(self):
         shutil.rmtree(self.tmpdir)

```